src/main/java/io/jenkins/plugins/opentelemetry/job/log/OtelLogSenderBuildListener.java

```diff
@@ -0,0 +1,199 @@
+     * configuration parameters transmitted via Jenkins remoting serialization
+     */
+    private static class OtelLogSenderBuildListenerOnAgent extends OtelLogSenderBuildListener {
+        private static final long serialVersionUID = 1;
+
+        private final static Logger logger = Logger.getLogger(OtelLogSenderBuildListenerOnAgent.class.getName());
+
+        /**
+         * Used to determine the clock adjustment on the Jenkins Agent.
+         */
+        private long instantInNanosOnJenkinsControllerBeforeSerialization;
+
+        /**
+         * Intended to be exclusively called on the Jenkins Controller by {@link OtelLogSenderBuildListenerOnController#writeReplace()}.
+         */
+        private OtelLogSenderBuildListenerOnAgent(@Nonnull BuildInfo buildInfo, @Nonnull Map<String, String> otelConfigProperties, @Nonnull Map<String, String> otelResourceAttributes) {
+            super(buildInfo, otelConfigProperties, otelResourceAttributes);
+            logger.log(Level.FINEST, () -> "new OtelLogSenderBuildListenerOnAgent()");
+            JenkinsJVM.checkJenkinsJVM();
+        }
+
+        /**
+         *
+         * @return
+         */
+        @Override
+        public LogEmitter getLogEmitter() {
+            JenkinsJVM.checkNotJenkinsJVM();
+            return GlobalOpenTelemetrySdk.getLogEmitter();
+        }
+
+        private void writeObject(ObjectOutputStream stream) throws IOException {
+            logger.log(Level.FINEST, () -> "writeObject(): set instantInNanosOnJenkinsControllerBeforeSerialization");
+            JenkinsJVM.checkJenkinsJVM();
+            this.instantInNanosOnJenkinsControllerBeforeSerialization = Clock.getDefault().now();
+            stream.defaultWriteObject();
+        }
+
+
+        private Object readResolve() {
+            adjustClock();
+            GlobalOpenTelemetrySdk.configure(
+                otelConfigProperties,
+                otelResourceAttributes,
+                true /* register shutdown hook when on the Jenkins agents */);
+            return this;
+        }
+
+        /**
+         * Timestamps of the logs emitted by the Jenkins Agents must be chronologically ordered with the timestamps of
+         * the logs & traces emitted on the Jenkins controller even if the system clock are not perfectly synchronized
+         */
+        private void adjustClock() {
+            JenkinsJVM.checkNotJenkinsJVM();
+            if (instantInNanosOnJenkinsControllerBeforeSerialization == 0) {
+                logger.log(Level.INFO, () -> "adjustClock(): unexpected timeBeforeSerialization of 0ns, don't adjust the clock");
+                this.clock = Clock.getDefault();
+            } else {
+                long instantInNanosOnJenkinsAgentAtDeserialization = Clock.getDefault().now();
+                long offsetInNanosOnJenkinsAgent = instantInNanosOnJenkinsControllerBeforeSerialization - instantInNanosOnJenkinsAgentAtDeserialization;
```

Comment:
Does this assume that the serialization, communication over Remoting, and deserialization are nearly instantaneous?

Reply:
>  I wanted monotonicity of the timestamps

My point: let us say the two hardware clocks are actually perfectly synchronized (up to relativistic constraints, whatever). But the Remoting channel is extremely sluggish today for whatever reason, and it takes ten seconds to send a `Command`. Then `offsetInNanosOnJenkinsAgent` will be 10_000_000_000 or so. But in fact the correct behavior in this case would be to use zero offset, if you are trying to ensure that messages delivered to the external storage are accurately threaded.

In other words, if you want to calculate clock skew, you need to send a _round-trip_ packet and adjust by half the time required to do so, or something a bit more complex like that. I think https://github.com/jenkinsci/jenkins/blob/38d4e86a80d123d50c966f4584a7e795b6bcb227/core/src/main/java/hudson/model/Slave.java#L732 does this.

> Don't you do something similar with the CloudWatch storage?

No, `io.jenkins.plugins.pipeline_cloudwatch_logs.TimestampTracker` solves different problems:

* ensuring that we do not mark the build log as complete until all messages known to have been sent from agents have actually arrived in the cloud and are available for reading (otherwise you would often lose the last few lines of a build, which tend to be the most critical if there was a failure)
* making sure output from multiple steps running _on a given agent_ is logically ordered (no attempt is made to order messages between the controller and an agent, or between two agents, in the face of clock skew)